File: src/orchestrator/discovery.py
"""
Service Discovery Registry.
Tracks service endpoints across the cluster for cross-node service resolution.

Services are registered by name and tracked across all nodes.
Each service has a list of endpoints (node_ip:port) where it can be reached.

DNS integration: Updates the DNS server configuration to resolve service names
to the appropriate node IPs.
"""

import logging
from typing import Dict, List, Optional
from datetime import datetime

logger = logging.getLogger(__name__)


class ServiceEndpoint:
    def __init__(self, node_name: str, node_address: str, container_id: str,
                 container_name: str, port: int = 0, status: str = "running"):
        self.node_name = node_name
        self.node_address = node_address  # IP:port of the node
        self.container_id = container_id
        self.container_name = container_name
        self.port = port
        self.status = status
        self.last_seen = datetime.utcnow().isoformat()

    def to_dict(self) -> dict:
        return {
            "node_name": self.node_name,
            "node_address": self.node_address,
            "container_id": self.container_id,
            "container_name": self.container_name,
            "port": self.port,
            "status": self.status,
            "last_seen": self.last_seen
        }


class ServiceRegistry:
    """In-memory service discovery registry backed by cluster state."""

    def __init__(self, cluster_state):
        self.cluster_state = cluster_state
        self._services: Dict[str, List[ServiceEndpoint]] = {}

    def refresh(self):
        """Rebuild service registry from cluster state placements."""
        self._services.clear()

        nodes = {n.name: n for n in self.cluster_state.list_nodes()}
        placements = self.cluster_state.get_placements()

        for p in placements:
            if not p.service_name or p.status != "running":
                continue

            node = nodes.get(p.node_name)
            if not node:
                continue

            endpoint = ServiceEndpoint(
                node_name=p.node_name,
                node_address=node.address,
                container_id=p.container_id,
                container_name=p.container_name,
                status=p.status
            )

            if p.service_name not in self._services:
                self._services[p.service_name] = []
            self._services[p.service_name].append(endpoint)

    def get_service(self, service_name: str) -> List[dict]:
        """Get all endpoints for a service."""
        self.refresh()
        endpoints = self._services.get(service_name, [])
        return [ep.to_dict() for ep in endpoints]

    def list_services(self) -> Dict[str, dict]:
        """List all services with endpoint counts."""
        self.refresh()
        result = {}
        for name, endpoints in self._services.items():
            running = [ep for ep in endpoints if ep.status == "running"]
            nodes = list(set(ep.node_name for ep in endpoints))
            result[name] = {
                "total_endpoints": len(endpoints),
                "running_endpoints": len(running),
                "nodes": nodes,
                "endpoints": [ep.to_dict() for ep in endpoints]
            }
        return result

    def get_service_nodes(self, service_name: str) -> List[str]:
        """Get list of node names where a service is running."""
        self.refresh()
        endpoints = self._services.get(service_name, [])
        return list(set(ep.node_name for ep in endpoints if ep.status == "running"))

    def generate_dns_config(self) -> str:
        """Generate dnsmasq-compatible DNS config for service discovery.

        Maps: {service_name}.svc.local -> node IPs
        """
        self.refresh()
        lines = ["# Auto-generated service discovery DNS config"]

        nodes = {n.name: n for n in self.cluster_state.list_nodes()}

        for service_name, endpoints in self._services.items():
            running = [ep for ep in endpoints if ep.status == "running"]
            seen_ips = set()
            for ep in running:
                # Extract IP from node address (address format: IP:port)
                ip = ep.node_address.split(":")[0] if ":" in ep.node_address else ep.node_address
                if ip not in seen_ips:
                    lines.append(f"address=/{service_name}.svc.local/{ip}")
                    seen_ips.add(ip)

        return "\n".join(lines)

File: internal/discovery/discovery.go
// Package discovery provides service discovery registry backed by cluster state.
// It tracks service endpoints across the cluster for cross-node service resolution
// and generates DNS configuration for dnsmasq-based service name resolution.
package discovery

import (
	"fmt"
	"strings"
	"time"

	"ai-container-go/internal/models"
)

// DiscoveryStateProvider is the interface for querying cluster state
// needed by the service registry.
type DiscoveryStateProvider interface {
	ListNodes(status ...models.NodeStatus) []models.NodeInfo
	GetPlacements(serviceName, nodeName string) []models.ContainerPlacement
}

// ServiceEndpoint represents a single reachable instance of a service.
type ServiceEndpoint struct {
	NodeName      string `json:"node_name"`
	NodeAddress   string `json:"node_address"`
	ContainerID   string `json:"container_id"`
	ContainerName string `json:"container_name"`
	Port          int    `json:"port"`
	Status        string `json:"status"`
	LastSeen      string `json:"last_seen"`
}

// ServiceRegistry is an in-memory service discovery registry that rebuilds
// its endpoint map from cluster state placements on every query.
type ServiceRegistry struct {
	state    DiscoveryStateProvider
	services map[string][]ServiceEndpoint
}

// NewServiceRegistry creates a ServiceRegistry backed by the given state provider.
func NewServiceRegistry(state DiscoveryStateProvider) *ServiceRegistry {
	return &ServiceRegistry{
		state:    state,
		services: make(map[string][]ServiceEndpoint),
	}
}

// Refresh rebuilds the service registry from cluster state placements.
// Only running placements that have a service name are included.
func (r *ServiceRegistry) Refresh() {
	r.services = make(map[string][]ServiceEndpoint)

	nodes := r.state.ListNodes()
	nodeMap := make(map[string]models.NodeInfo, len(nodes))
	for _, n := range nodes {
		nodeMap[n.Name] = n
	}

	placements := r.state.GetPlacements("", "")

	for _, p := range placements {
		if p.ServiceName == "" || p.Status != "running" {
			continue
		}

		node, ok := nodeMap[p.NodeName]
		if !ok {
			continue
		}

		ep := ServiceEndpoint{
			NodeName:      p.NodeName,
			NodeAddress:   node.Address,
			ContainerID:   p.ContainerID,
			ContainerName: p.ContainerName,
			Port:          0,
			Status:        p.Status,
			LastSeen:      time.Now().UTC().Format(time.RFC3339),
		}

		r.services[p.ServiceName] = append(r.services[p.ServiceName], ep)
	}
}

// GetService returns all endpoints for the named service.
func (r *ServiceRegistry) GetService(serviceName string) []ServiceEndpoint {
	r.Refresh()
	eps, ok := r.services[serviceName]
	if !ok {
		return []ServiceEndpoint{}
	}
	return eps
}

// ListServices returns a summary for every registered service.
// Each entry contains total_endpoints, running_endpoints, nodes, and endpoints.
func (r *ServiceRegistry) ListServices() map[string]map[string]any {
	r.Refresh()

	result := make(map[string]map[string]any, len(r.services))
	for name, endpoints := range r.services {
		running := 0
		nodeSet := make(map[string]struct{})
		for _, ep := range endpoints {
			if ep.Status == "running" {
				running++
			}
			nodeSet[ep.NodeName] = struct{}{}
		}

		nodes := make([]string, 0, len(nodeSet))
		for n := range nodeSet {
			nodes = append(nodes, n)
		}

		result[name] = map[string]any{
			"total_endpoints":   len(endpoints),
			"running_endpoints": running,
			"nodes":             nodes,
			"endpoints":         endpoints,
		}
	}
	return result
}

// GetServiceNodes returns the unique node names where the named service
// has running endpoints.
func (r *ServiceRegistry) GetServiceNodes(serviceName string) []string {
	r.Refresh()

	endpoints := r.services[serviceName]
	nodeSet := make(map[string]struct{})
	for _, ep := range endpoints {
		if ep.Status == "running" {
			nodeSet[ep.NodeName] = struct{}{}
		}
	}

	nodes := make([]string, 0, len(nodeSet))
	for n := range nodeSet {
		nodes = append(nodes, n)
	}
	return nodes
}

// GenerateDNSConfig produces a dnsmasq-compatible DNS configuration string
// that maps each service to its node IPs under the .svc.local domain.
func (r *ServiceRegistry) GenerateDNSConfig() string {
	r.Refresh()

	lines := []string{"# Auto-generated service discovery DNS config"}

	for serviceName, endpoints := range r.services {
		seenIPs := make(map[string]struct{})
		for _, ep := range endpoints {
			if ep.Status != "running" {
				continue
			}
			ip := ep.NodeAddress
			if idx := strings.Index(ip, ":"); idx != -1 {
				ip = ip[:idx]
			}
			if _, seen := seenIPs[ip]; !seen {
				lines = append(lines, fmt.Sprintf("address=/%s.svc.local/%s", serviceName, ip))
				seenIPs[ip] = struct{}{}
			}
		}
	}

	return strings.Join(lines, "\n")
}
